### ISSUE TITLE
Type spec for google.protobuf.Any conflicts with any type

### ISSUE THREAD

**Issue by ostera** [tomas-abrahamsson/gpb#165]
Hello!

I was using grpcbox by @tsloughter and ended up filing an issue for being able to rename a type-spec: https://github.com/tsloughter/grpcbox/issues/13

It seems that the well known protofiles are included in this repo, so I'm wondering how are you dealing with the `any` type spec rename?

In particular I'm seeing:

```erl
%... part of the generated code
-type any() ::
     #{type_url                => iodata(),        % = 1
       value                   => iodata()         % = 2
      }.
```

Which triggers a dyalizer error.

I have temporarily renamed the message to `Msg` so it doesn't block me during development until I find a better way of dealing with this.

Thanks for your work on `gpb`! 🙌

**Comment by tomas-abrahamsson** [comment]
Oh, I see. It never occurred to me that there could be collisions, but of course there can. I guess one way to go could be to invent type-prefix and -suffix [`rename`ing option](https://hexdocs.pm/gpb/gpb_compile.html#option-rename). (See the [`renaming` type](https://hexdocs.pm/gpb/gpb_compile.html#type-renaming) for a brief overview of what renamings exist.) Does this sound reasonable? If so, I could look into implementing it.

What exists today is of course the [`{type_specs, false}`](https://hexdocs.pm/gpb/gpb_compile.html#option-type_specs) option as @tsloughter mentioned.

(Along the lines of what exist today, possibly, the [`any_translate`](https://hexdocs.pm/gpb/gpb_compile.html#option-any_translate) or more generally the [`translate_type`](https://hexdocs.pm/gpb/gpb_compile.html#option-translate_type) options could be of some use. They make it possible to specify packers and unpackers to a different internal type, for instance an `uint32` could be translated to and from an ip-address. Part of this is also that it is possible to specify a different type spec for the translated type, but if I remember correctly, it might be impossible to just specify a typespec, it may be that one must also specify packer/unpacker functions. It would of course be possible to specify them as `id(X) -> X`, but it would make this solution a little clumsy. See the [`translation()`](https://hexdocs.pm/gpb/gpb_compile.html#type-translation) type for some more details.)

But all in all, I think the cleanest way (for a user) for be to add for instance a `{rename,{msg_typename,{prefix,"msg_"}}}` option. The `msg_typename` part would be new.

**Comment by ostera** [comment]
@tomas-abrahamsson I think for now the `{rename, {msg_name, {prefix, "msg_"}}}` would be good. Is this something I could help implement? :) Happy to do so if you could point me in the right direction.

**Comment by tomas-abrahamsson** [comment]
> I think for now the {rename, {msg_name, {prefix, "msg_"}}} would be good. ... Happy to do so if you could point me in the right direction.

Absolutely! Assuming you meant a `msg_typename` option¹ (?) I think the way to go would be to compute a `TypeRenamings` and add it in a `{types, TypeRenamings}` item [in gpb_names.erl](https://github.com/tomas-abrahamsson/gpb/blob/4.7.2/src/gpb_names.erl#L285-L293). Then pick it up and use it from the [`AnRes#anres.renamings` field](https://github.com/tomas-abrahamsson/gpb/blob/4.7.2/src/gpb_compile.hrl#L21) at various places in [gpb_gen_types.erl](https://github.com/tomas-abrahamsson/gpb/blob/4.7.2/src/gpb_gen_types.erl) for example [here](https://github.com/tomas-abrahamsson/gpb/blob/4.7.2/src/gpb_gen_types.erl#L93), [here](https://github.com/tomas-abrahamsson/gpb/blob/4.7.2/src/gpb_gen_types.erl#L99) and [here](https://github.com/tomas-abrahamsson/gpb/blob/4.7.2/src/gpb_gen_types.erl#L438-L439)  (you may need to pass on the `AnRes` or only `TypeRenamings` to sub functions). The `TypeRenamings` should probably be a [dict()](http://erlang.org/doc/man/dict.html) from `MsgName::atom() -> TypeName::atom()`. I've stayed with dict instead of maps due to support still for Erlang pre-17, but #133 _Drop Erlang R16 support_ is soon due since Erlang 22 is about to be released soon.

¹ However, if you really meant the `msg_name` renaming option, then I just realized this could possibly be a quick and easy fix for your problem: renaming options are stackable. And if I remember correctly, the grpcbox already sets renaming options for `snake_case` and `basename`, if it allows you to specify an extra option, the `{rename,{msg_name,{prefix,"msg_"}}}` you mentioned, then all your protobuf message names would begin with `msg_`, so the `any` would become `msg_any` and its type would no longer collide with the built-in. All message names would change, but since you are using maps (as opposed to records), I think the only impact would be that (a) type names start with `msg_` and (b) if you would ever need to manually encode or decode a message, you would have to just prepend `msg_` in call like for example `⟨generated-code⟩:encode(MyMap, msg_my_message)`. If this is acceptable (and works) then it may be the quick fix, since the `{rename,{msg_name,{prefix,"msg_"}}}` already exists and works.

Hope this rambling can make some sense to you, feel free to ask if I'm unclear.

**Comment by ostera** [comment]
Oh, I actually had a typo on my message 😂. I will have a look at renaming the message, but I'm curious if you are aware of any serialization/deserialization issues that may come from changing it?

If there are, I'll look into contributing the `{msg_typename}` option.

**Comment by tomas-abrahamsson** [comment]
Over the wire only the field number is sent, no field names and no message names, so it should not cause serialization issues. 

I'm not sure if there will be any issues with grpcbox with mapping from external rpc measage names to internal, or if this is all handled automatically with the introspection functions now. @tsloughter would know, I suppose?

**Comment by tomas-abrahamsson** [comment]
Just following up, did you manage to get around using the already existing `{rename,{msg_name,{prefix,"msg_"}}}` option, or are you looking into (or still need) adding possibilities for renaming `msg_typename`s?

**Comment by tomas-abrahamsson** [comment]
@ostera, I made an attempt at the renaming options. On the (temporary) branch `rename-types`, I added support for the options `{rename,{msg_typename,How}}` and `{rename,{enum_typename,How}}` where `How` can be for instance `{prefix,atom() | string()}` as discussed earlier. Is there still a need for this?

**Comment by tomas-abrahamsson** [comment]
I decided to merge the `rename-types` branch after all, it is included in 4.9.0, which I just pushed.
